repos/system_upgrade/el8toel9/actors/firewalldcollectusedobjectnames/libraries/private_firewalldcollectusedobjectnames.py

```diff
@@ -0,0 +1,90 @@
+from leapp.models import FirewalldUsedObjectNames
+
+try:
+    from firewall.core.fw import Firewall
+except ImportError:
+    pass
+
+
+def is_zone_in_use(conf):
+    # This does not account for zone assignments given by NetworkManager.
+    if conf.get('interfaces', []) or conf.get('sources', []):
+        return True
+
+    return False
+
+
+def is_policy_in_use(conf, used_zones):
+    # A policy is in use if both ingress_zones and egress_zones contain at
+    # least one of following: an active zone, 'ANY', 'HOST'.
+    for zone in conf.get('ingress_zones', []):
+        if zone in ['ANY', 'HOST'] or zone in used_zones:
+            break
+    else:
+        return False
+    for zone in conf.get('egress_zones', []):
+        if zone in ['ANY', 'HOST'] or zone in used_zones:
+            return True
+    return False
+
+
+def get_used_services(conf, isZone):
+    used_services = set()
+
+    for service in conf.get('services', []):
+        used_services.add(service)
+
+    # Also need to look for 'service name="<service>"' in rich rules. The rule
+    # strings are normalized by firewalld. zone keyword is 'rules_str'. policy
+    # keyword is 'rich_rules'.
+    for rule in conf.get('rules_str', []) if isZone else conf.get('rich_rules', []):
+        try:
+            search = 'service name="'
+            start = rule.index(search) + len(search)
+            stop = rule[start:].index('"')
+            used_services.add(rule[start:start+stop])
+        except ValueError:
+            pass
+
+    return used_services
+
+
+def read_config():
+    if 'Firewall' not in dir():
+        # import failure missing means firewalld is not installed. Just return
+        # the defaults.
+        return FirewalldUsedObjectNames()
```

Comment:
@Rezney, this if statement is evaluating to True when running `leapp preupgrade --debug --no-rhsm`. It works as expected inside a repl.

The only reason the import check is like this is because a local import was denied by CI. I had to put the import at the top of the file.

Inside the if statement sys.path = `['/usr/bin', '/usr/lib6/python36.zip', '/usr/lib64/python3.6', '/usr/lib64/python3.6/lib-dynload', '/usr/lib64/python3.6/site-packages', '/usr/lib/python3.6/site-packages']

Reply:
Since `dir()` is inside a function `Firewall` is in the global scope and not returned by `dir()`.

Solution is to try and use `Firewall()` and catch the exception.

Reply:
Done. Please take a look.